src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,702 @@
+  // First, un-permute the group to get back to the canonical order, offset by 1
+  // if we are newly inserting a brand.
+  std::vector<HeapType> canonical(info.group.size() + insertingBrand);
+  for (size_t i = 0; i < info.group.size(); ++i) {
+    canonical[info.permutation[i] + insertingBrand] = info.group[i];
+  }
+  // Update the brand.
+  if (brand) {
+    canonical[0] = **brand;
+  }
+  if (insertingBrand) {
+    info.group.resize(info.group.size() + 1);
+    info.hasBrand = true;
+  }
+  // Finally, re-permute with the new permutation..
+  info.permutation = *orders;
+  for (size_t i = 0; i < info.group.size(); ++i) {
+    info.group[i] = canonical[info.permutation[i]];
+  }
+}
+
+std::array<Field, BrandTypeIterator::optionCount>
+  BrandTypeIterator::fieldOptions = {{}};
+
+void BrandTypeIterator::initFieldOptions() {
+  BrandTypeIterator::fieldOptions = {{
+    Field(Field::i8, Mutable),
+    Field(Field::i16, Mutable),
+    Field(Type::i32, Mutable),
+    Field(Type::i64, Mutable),
+    Field(Type::f32, Mutable),
+    Field(Type::f64, Mutable),
+    Field(Type(HeapType::any, Nullable), Mutable),
+    Field(Type(HeapType::func, Nullable), Mutable),
+    Field(Type(HeapType::ext, Nullable), Mutable),
+    Field(Type(HeapType::none, Nullable), Mutable),
+    Field(Type(HeapType::nofunc, Nullable), Mutable),
+    Field(Type(HeapType::noext, Nullable), Mutable),
+    Field(Type(HeapType::any, NonNullable), Mutable),
+    Field(Type(HeapType::func, NonNullable), Mutable),
+    Field(Type(HeapType::ext, NonNullable), Mutable),
+    Field(Type(HeapType::none, NonNullable), Mutable),
+    Field(Type(HeapType::nofunc, NonNullable), Mutable),
+    Field(Type(HeapType::noext, NonNullable), Mutable),
+  }};
+}
+
+struct MinimizeRecGroups : Pass {
+  // The types we are optimizing and their indices in this list.
+  std::vector<HeapType> types;
+
+  // A global ordering on all types, including public types. Used to define a
+  // total ordering on rec group shapes.
+  std::unordered_map<HeapType, size_t> typeIndices;
+
+  // As we process strongly connected components, we will construct output
+  // recursion groups here.
+  std::vector<RecGroupInfo> groups;
+
+  // For each shape of rec group we have created, its index in `groups`.
+  std::unordered_map<RecGroupShape, size_t> groupShapeIndices;
+
+  // When we find that two groups are isomorphic to (i.e. permutations of)
+  // each other, we combine their equivalence classes and choose a new class
+  // representative to use to disambiguate the groups.
+  DisjointSets equivalenceClasses;
+
+  void run(Module* module) override {
+    // There are no recursion groups to minimize if GC is not enabled.
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    initBrandOptions();
+
+    types = ModuleUtils::getPrivateHeapTypes(*module);
+    for (auto type : ModuleUtils::collectHeapTypes(*module)) {
+      typeIndices.insert({type, typeIndices.size()});
+    }
```

Comment:
It looks like `types` is only the private types, while `typeIndices` is all the types. That could be confusing - perhaps rename the former `privateTypes`?

Reply:
Will keep that in mind when I fix this problem with public types.